tar_scm.py

```diff
@@ -72,6 +72,17 @@ def fetch_upstream_git(url, clone_dir, revision, cwd, kwargs):
     """Fetch sources via git."""
     safe_run(['git', 'clone', url, clone_dir], cwd=cwd,
              interactive=sys.stdout.isatty())
+    if revision:
+        # check if the reference already exists.
+        try:
+            safe_run(['git', 'rev-parse', '--verify', '--quiet', revision],
+                     cwd=clone_dir, interactive=sys.stdout.isatty())
+        except SystemExit:
+            # fetch reference from url and create locally
+            safe_run(['git', 'fetch', url, revision], cwd=clone_dir,
```

Comment:
If you change this to

    ['git', 'fetch', url, revision + ':' + revision]

then you can eliminate the `update-ref` command below.